### PR TITLE
Follow-up to comments on #6654 (support for `autoBuild` and `deployByDefault`)

### PR DESCRIPTION
**Pull request by rm3l**
**What type of PR is this:**
/kind code-refactoring

**What does this PR do / why we need it:**
This is a follow-up PR to https://github.com/redhat-developer/odo/pull/6654 (support for `autoBuild` and `deployByDefault` fields), addressing some of the comments left there.

**Which issue(s) this PR fixes:**
Related to #5694 

**PR acceptance criteria:**

- [ ] Unit test 

- [x] Integration test 

- [x] [Documentation](https://deploy-preview-6720--odo-docusaurus-preview.netlify.app/docs/development/devfile#how-odo-determines-components-that-are-applied-automatically) 

**How to test changes / Special notes to the reviewer:**

### CODE REVIEW

File: tests/examples/source/devfiles/nodejs/devfile-autobuild-deploybydefault.yaml [hunk 1]
```diff
@@ -37,7 +37,7 @@ commands:
 
   - composite:
       commands:
-        - image-autobuild-true-and-referenced
```
Comment (valaparthvi) [inline tests/examples/source/devfiles/nodejs/devfile-autobuild-deploybydefault.yaml:40]: Perhaps what you had makes sense while reading it. Can you revert this change?

**Review by valaparthvi — CHANGES_REQUESTED** [review]
Thank you for following up on these!

File: tests/examples/source/devfiles/nodejs/devfile-autobuild-deploybydefault.yaml (hunk 1 above)
Comment (rm3l) [reply tests/examples/source/devfiles/nodejs/devfile-autobuild-deploybydefault.yaml]: Indeed. See [`4bdcec6` (#6720)](https://github.com/redhat-developer/odo/pull/6720/commits/4bdcec60089362915b518c41b85929532faddc53)

**Comment by feloy** [comment]
```
 [FAILED] [12.944 seconds]
odo dev debug command tests when starting with Devfile with autoBuild or deployByDefault components when running odo dev with some components not referenced in the Devfile [BeforeEach] should create the appropriate resources
  [BeforeEach] /go/odo_1/tests/integration/cmd_dev_debug_test.go:429
  [It] /go/odo_1/tests/integration/cmd_dev_debug_test.go:459
[...]
  [odo]  âœ—  failed to create Kubernetes resources associated with the component: pods "ocp-deploybydefault-true-and-not-referenced" is forbidden: error looking up service account cmd-dev-debug-test459zaf/default: serviceaccount "default" not found

```

**Comment by feloy** [comment]
/override windows-integration-test/Windows-test

Flaky e2e test
```
[FAILED] [206.998 seconds]
E2E Test starting with non-empty Directory add Binding [It] should verify developer workflow of using binding as env in innerloop
C:/Users/Administrator.ANSIBLE-TEST-VS/3652/tests/e2escenarios/e2e_test.go:337
```
